refactor: move limit query from state to urlParams

OgDev-01 opened the pull request:
## Description
This PR refactors how the limit query state is stored across the application. Now we have the limit in the url of the pages
<!-- 
Please do not leave this blank 
This PR [adds/removes/fixes/replaces] the [feature/bug/etc]. 
-->

## What type of PR is this? (check all applicable)

- [ ] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [x] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents
closes #2154
<!-- 
Please use this format link issue numbers: Fixes #123
https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword 
-->

## Mobile & Desktop Screenshots/Recordings

<!-- Visual changes require screenshots -->


## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [x] 🙅 no documentation needed

## [optional] Are there any post-deployment tasks we need to perform?



## [optional] What gif best describes this PR or how it makes you feel?

<img src="https://media0.giphy.com/media/3oEjHI8WJv4x6UPDB6/giphy.gif"/>

<!-- note: PRs with deleted sections will be marked invalid -->

<!--
  For Work In Progress Pull Requests, please use the Draft PR feature,
  see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.
  
  For a timely review/response, please avoid force-pushing additional
  commits if your PR already received reviews or comments.
  
  Before submitting a Pull Request, please ensure you've done the following:
  - 📖 Read the Open Sauced Contributing Guide: https://github.com/open-sauced/.github/blob/main/CONTRIBUTING.md.
  - 📖 Read the Open Sauced Code of Conduct: https://github.com/open-sauced/.github/blob/main/CODE_OF_CONDUCT.md.
  - 👷‍♀️ Create small PRs. In most cases, this will be possible.
  - ✅ Provide tests for your changes.
  - 📝 Use descriptive commit messages.
  - 📗 Update any related documentation and include any relevant screenshots.
-->

nickytonline commented on lib/hooks/api/useContributorList.ts:
```diff
@@ -30,16 +30,15 @@ export const useContributorsList = ({
+  defaultLimit?: string;
   defaultRange?: string;
 }) => {
-  const [range, setRange] = useState(defaultRange); // [start, end
```
Refactor (nit): Prefer a separate pull request for the changes for range as it is not related to the current refactor.

nickytonline commented on lib/hooks/api/useContributorList.ts:
```diff
@@ -30,16 +30,15 @@ export const useContributorsList = ({
   query.append("page", page.toString());
-  query.append("limit", limit.toString());
-  query.append("range", range);
+  query.append("limit", defaultLimit ?? "10");
```
Refactor (blocking): The `defaultLimit` already has a default value of `"10"`, so `?? "10"` is note required.

```suggestion
  query.append("limit", defaultLimit);
```

nickytonline commented on pages/lists/[listId]/activity.tsx:
```diff
@@ -31,16 +31,23 @@ interface ContributorListPageProps {
   isOwner: boolean;
 }
 
+export type ServerFilterParams = {
```
Naming (nit): No need to specify `Server` in the type name.

```suggestion
export type FilterParams = {
```

Also, `FilterParams` is used on the list activity page but also elsewhere. Consider moving the type elsewhere and import it into the list activity page.

nickytonline commented on pages/lists/[listId]/activity.tsx:
```diff
@@ -31,16 +31,23 @@ interface ContributorListPageProps {
   } = await supabase.auth.getSession();
   const bearerToken = session ? session.access_token : "";
-  const { listId, range: rawRange } = ctx.params as { listId: string; range: string };
+  const { listId, range: rawRange, limit: RawLimit } = ctx.params as ServerFilterParams;
```
Refactor (blocking): `RawLimit` should be camel case as it's a JavaScript variable and not a type.

```suggestion
  const { listId, range: rawRange, limit: rawLimit } = ctx.params as ServerFilterParams;
```

nickytonline reviewed (changes requested):
This is looking good @OgDev-01, just some change requests.

brandonroberts commented:
@OgDev-01 I think the changes are good, but using strings as something that should be a number seems off. I know that's how they are in the URL, but I think we should convert them to a numbers at some point.

nickytonline commented:
> @OgDev-01 I think the changes are good, but using strings as something that should be a number seems off. I know that's how they are in the URL, but I think we should convert them to a numbers at some point.

Yeah it seemed off to me too. I mentioned it in the changes for the range changes PR as well although it wasn't blocking the PR. Agree with Brandon here even if they become strings in the end.

We can revisit it for `range` after.

jpmcb commented on components/atoms/Select/limit-select.tsx:
```diff
@@ -6,13 +6,17 @@ import { Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from "c
-  const [selected, setSelected] = useState<null | string>(null);
-  const handleSelected = (value: string) => {
+
+export type limitSelectMap = "10" | "20" | "30" | "40" | "50";
```
@nickytonline there are a few places within the API where we default to some set limit. Example:

https://github.com/open-sauced/api/blob/24892059033ee8e9855517d887475cdb54f8681f/src/user/dtos/top-users.dto.ts#L25-L36

But as far as I can tell, these numbers are arbitrary. The upper limit on many of these are intended to prevent full table scans.

jpmcb commented on pages/lists/[listId]/contributors.tsx:
```diff
@@ -69,10 +70,13 @@ const ContributorsListPage = ({ list, initialData, isError, isOwner }: Contribut
+  } = useContributorsList({
+    listId: list?.id,
+    initialData,
+    defaultRange: range as string,
+    defaultLimit: limit as string,
```
Looks like do a cast here we cast this number to a string to be consumed in the URL? Like brandon mentioned, I'm also curious if there's a way to prevent storing strings as numbers (especially if it'll be converted to a number eventually and cast back to a string here)

nickytonline replied:
This one is king of inevitable because it comes from the router via the querystring. It's type is `string | string[] | undefined`, so in this case, this would need to become most likely this

`defaultLimit: Number(limit as string),`

OgDev-01 commented:
Hey @open-sauced/engineering, I've reverted the limit type to number and converted it to string in the appropriate areas. this is ready for a second review 🍕
